### PR TITLE
[HLAPI] Support advanced RSQL syntax

### PR DESCRIPTION
**Pull request by cconard96**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

Currently, the RSQL filter feature only supports separating filters by `;` which is treated like an 'AND' operator.
This PR expands functionality by adding condition group support (parenthesis) and logical 'OR' support via the `,` separator.
This should mean that RSQL filtering is equivalent with the search capabilities of the regular search engine used in the web and legacy API.

### CODE REVIEW

**Review by orthagh — APPROVED** [review]

File: src/Api/HL/RSQLInput.php [hunk 1]
```diff
@@ -78,186 +78,266 @@ public function getOperators(): array
...
+        // RSQL filters may be grouped with parenthesis
+        // Filters separated by a comma are OR'd together
+        // Filters separated by a semicolon are AND'd together
+        // When there is a mix of commas and semicolons within a group, the semicolons take precedence
+
+        // Parse the RSQL filter string into an array similar to the DBIterator criteria syntax
+        // For example, the RSQL filter string: "name==test1;name==test2,name==test3" would be parsed into:
+        // [0 => ['name==test1'], 'OR' => ['name==test2', 'name==test3']]
+        // The string: ("name==test1;name==test2),name==test3" would be parsed into:
+        // [0 => ['name==test1', 'name==test2'], 'OR' => ['name==test3']]
```
Comment (cedric-anne) [inline src/Api/HL/RSQLInput.php:218]: `(name==test1 AND nam==test2) OR name==test3` have to be passed this way to `DBmysql::request()`:
```php
$where = [
    'OR' => [
        [
            ['name' => 'test1'],
            ['name' => 'test2'],
        ],
        ['name' => 'test3']
    ]
];
```

```suggestion
        // The string: "(name==test1;name==test2),name==test3" would be parsed into:
        // ['OR' => [['name==test1', 'name==test2'], ['name==test3']]]
```

File: src/Api/HL/RSQLInput.php [hunk 2]
```diff
@@ -78,186 +78,266 @@ public function getOperators(): array
...
+        }
+        return $operators;
     }
 
-    /**
-     * @return array
-     * @phpstan-return {field: string, value: string, callable: callable}[]
-     */
-    private function getFilters()
+    public function getGroupedFilters(string $filter_string, array $flattened_props)
```
Comment (cedric-anne) [inline src/Api/HL/RSQLInput.php:207]: It is pretty hard to read. Could you add a test on this method, with complex cases?

**Comment by cconard96** [comment]
Regex matching to parse RSQL queries proved too complex. It wasn't handling nested filter groups and a few other cases.
PR was updated to replace the `RSQLInput` class with a `Lexer` class which scans the query and tokenizes it, and a `Parser` class which then takes the tokens and builds SQL criteria.
Another change is that the final SQL criteria is a built SQL string inside a `QueryExpression` rather than an array of criteria. It seems too complex to build the array and unnecessary. Luckily, SQL databases process the logical AND/OR operators the same as is expected with RSQL (AND first).

Finally, since the RSQL -> SQL steps are more clearly defined, it was easier to add user-friendly error messages when an invalid query is specified.

Tests have been added.

File: src/Api/HL/RSQL/Parser.php [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+
+    public function getOperators(): array
+    {
+        static $operators = null;
+
+        if ($operators === null) {
+            $operators = [
+                [
+                    'operator' => '==',
+                    'description' => 'equivilent to',
```
Comment (cedric-anne) [inline src/Api/HL/RSQL/Parser.php:92]:
```suggestion
                    'description' => 'equivalent to',
```

File: src/Api/HL/RSQL/Parser.php [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+                    'sql_where_callable' => fn ($a, $b) => [
+                        [$this->db::quoteName($a) => $b]
+                    ],
+                ],
+                [
+                    'operator' => '!=',
+                    'description' => 'not equivilent to',
```
Comment (cedric-anne) [inline src/Api/HL/RSQL/Parser.php:99]:
```suggestion
                    'description' => 'not equivalent to',
```

**Review by cedric-anne — APPROVED** [review]

**Review by orthagh — APPROVED** [review]
still good for me

**Pull request merged**